src/Compiler/Interpreter/VMCode.idr

```diff
@@ -0,0 +1,281 @@
+ioRes obj = Constructor (Left 0) [Const WorldVal, obj]
+
+-- TODO: add more?
+knownForeign : NameMap (ar ** (Ref State InterpState => Stack -> Vect ar Object -> Core Object))
+knownForeign = fromList
+    [ (NS_UN ioNS "prim__putChar", (2 ** prim_putChar))
+    , (NS_UN ioNS "prim__getChar", (1 ** prim_getChar))
+    , (NS_UN ioNS "prim__getStr", (1 ** prim_getStr))
+    , (NS_UN ioNS "prim__putStr", (2 ** prim_putStr))
+    ]
+  where
+    prim_putChar : Ref State InterpState => Stack -> Vect 2 Object -> Core Object
+    prim_putChar _ [Const (Ch c), Const WorldVal] = ioRes unit <$ coreLift_ (putChar c)
+    prim_putChar stk as = argError stk as
+    prim_getChar : Ref State InterpState => Stack -> Vect 1 Object -> Core Object
+    prim_getChar _ [Const WorldVal] = ioRes . Const . Ch <$> coreLift getChar
+    prim_getChar stk as = argError stk as
+    prim_getStr : Ref State InterpState => Stack -> Vect 1 Object -> Core Object
+    prim_getStr _ [Const WorldVal] = Const . Str <$> coreLift getLine
+    prim_getStr stk as = argError stk as
+    prim_putStr : Ref State InterpState => Stack -> Vect 2 Object -> Core Object
+    prim_putStr _ [Const (Str s), Const WorldVal] = ioRes unit <$ coreLift_ (putStr s)
+    prim_putStr stk as = argError stk as
+
+knownExtern : NameMap (ar ** (Ref State InterpState => Stack -> Vect ar Object -> Core Object))
+knownExtern = empty
+
+beginFunction : Ref State InterpState => List (Int, Object) -> List VMInst -> Int -> Core (List VMInst)
+beginFunction args (DECLARE (Loc i) :: is) maxLoc = beginFunction args is (Prelude.max i maxLoc)
+beginFunction args (DECLARE _ :: is) maxLoc = beginFunction args is maxLoc
+beginFunction args (START :: is) maxLoc = do
+    locals <- coreLift $ newArray (maxLoc + 1)
+    ignore $ traverse (\(idx, arg) => coreLift $ writeArray locals idx arg) args
+    update State $ record { locals = locals, return = Nothing }
+    pure is
+beginFunction args is maxLoc = pure is
+
+parameters {auto c : Ref Ctxt Defs}
+  mutual
+    step : Stack -> Ref State InterpState => VMInst -> Core ()
+    step stk (DECLARE _) = pure ()
+    step stk START = pure ()
+    step stk (ASSIGN target val) = do
+        valObj <- getReg stk val
+        setReg stk target valObj
+    step stk (MKCON target tag args) = do
+        argObjs <- traverse (getReg stk) args
+        setReg stk target $ Constructor tag argObjs
+    step stk (MKCLOSURE target fn missing args) = do
+        argObjs <- traverse (getReg stk) args
+        setReg stk target $ Closure (pred missing) argObjs fn
+    step stk (MKCONSTANT target c) = setReg stk target $ Const c
+    step stk (APPLY target fn arg) = do
+        fnObj <- getReg stk fn
+        argObj <- getReg stk arg
+        case fnObj of
+            Closure Z args fn => do
+                res <- callFunc stk fn (args ++ [argObj])
+                setReg stk target res
+            Closure (S k) args fn => setReg stk target $ Closure k (args ++ [argObj]) fn
```

Comment:
Should `Closure` store its list of args using a snoc list? You'd have:

```idris
callFunc stk fn (args <>> [argObj])
setReg stk target $ Closure k (args :< argObj)
```

`(<>>)` is still linear in `args` but `(:<)` is constant time!